### ISSUE TITLE
FR: auth to hosted KFP endpoint using google application default credentials

### ISSUE THREAD

**Issue by amygdala** [kubeflow/pipelines#3793]
Currently, to connect to the hosted kfp endpoint, it's required that gcloud is installed, or else that a token is passed to the KFP client. 
Auth doesn't work via just setting
`export GOOGLE_APPLICATION_CREDENTIALS=/path/to/<gce_service_account>.json`, but it would be very useful if it did.

Note: I thought the following might work (after setting the `GOOGLE_APPLICATION_CREDENTIALS` var) , but it doesn't...

```python
from oauth2client.client import GoogleCredentials
credentials = GoogleCredentials.get_application_default()
credentials.get_access_token()
token1 = credentials.access_token
client=kfp.Client(host='xxxxx-dot-us-central2.pipelines.googleusercontent.com', existing_token=token1)
```

/cc @rmgogogo

**Comment by numerology** [comment]
Hi @amygdala 

Actually I thought your code snippet should work... What's the error did you get?

**Comment by amygdala** [comment]
There were no connection errors, but the response looked like this:
```
client.list_pipelines()
{'next_page_token': None, 'pipelines': None, 'total_size': None}
```

(I double checked that I did not see the same when testing against the same endpoint, but not setting the token, on a machine with gcloud installed.  This was just to make sure there wasn't something weird about my KFP installation. But with using gcloud, I got the expected content returned).

I wonder if I somehow did something wrong.  It is probably worth your trying to repro if the code looks like it should work.

**Comment by amygdala** [comment]
Btw, I wondered if maybe the GCE service account need to also have 'service account token creator' permissions.  I added that permission, but it didn't seem to make a difference. (though again... it's possible I made a mistake somewhere).

**Comment by numerology** [comment]
Thanks @amygdala I'll try to reproduce on my end.

In the meantime, can you confirm you granted the project editor role to the GCE SA you're using? The all-none response issue is very likely to be caused by insufficient permission.

**Comment by amygdala** [comment]
Yes, it all looks okay (I tried both before and after adding the token creator privs, but it always was a project editor):

![image](https://user-images.githubusercontent.com/115093/82501933-f2fce800-9aaa-11ea-99bb-defcfeea35c0.png)

As far as I know I didn't do anything goofy, so am interested to know if you can repro.

**Comment by amygdala** [comment]
do you think maybe it needs to be project owner?

**Comment by rmgogogo** [comment]
GCE SA should already have all permissions. Back to you a little bit later after reproduce the issue.
(I may try use GCESA's ADC in a VM)

**Comment by rmgogogo** [comment]
```
from oauth2client.client import GoogleCredentials
import kfp

credentials = GoogleCredentials.get_application_default()
credentials.get_access_token()
token = credentials.access_token

client=kfp.Client(host='39e0f88a26ceb5d1-dot-us-central2.pipelines.googleusercontent.com', existing_token=token)
client.list_pipelines()
```

I can well fetch out the pipelines via token from GOOGLE_APPLICATION_CREDENTIALS.
Amy, do you still can hit the problem?

Did you setup the `export GOOGLE_APPLICATION_CREDENTIALS=xxx` correctly and the credential's service account can access the KFP? (e.x. same project's GCE default service account).

**Comment by blazickjoe** [comment]
I've seen this issue when mistakenly adding and ending "/" to the host. May want to check that as well.

**Comment by aaaaahaaaaa** [comment]
@rmgogogo 
Your code doesn't work for me. I'm getting back an empty payload:
```
{'next_page_token': None, 'pipelines': None, 'total_size': None}
```

Where I'm confused is that `credentials.access_token` returns a JWT, which doesn't seem to be the expected format. If I'm using the token generated with `gcloud auth print-access-token` (not `gcloud auth print-identity-token`) as the `existing_token` parameter, then it works.

I'm not sure what format is `gcloud auth print-access-token`, it doesn't look like at a JWT.

**Comment by amygdala** [comment]
Hmm.  Yes, I _thought_ I set up everything as described.  I think to hit the error you have to test w/out having gcloud installed (though not completely sure).  @rmgogogo 
I will try again in a day or so but since there's a repro from someone else I'll unassign me if that's okay.

**Comment by rmgogogo** [comment]
I'm wondering whether it's because I test the pieces of codes in the VM belongs to the same project.
Let me try again in another project or local machine to see if it fails. Back to here later.

**Comment by rmgogogo** [comment]
I also can reproduce the issue.
In a VM with the same project. credentials.get_access_token() gave me the correct value (a shorter string). The type of the credentials is "oauth2client.contrib.gce.AppAssertionCredentials".

In a local Linux, it outputs JWT and list result is empty. The type of the credentials is "oauth2client.service_account._JWTAccessCredentials"

**Comment by rmgogogo** [comment]
https://google-auth.readthedocs.io/en/latest/oauth2client-deprecation.html

It's suggested to use 
```
from google.oauth2 import service_account
credentials = service_account.Credentials.from_service_account_file("/usr/local/google/home/renming/adc.json", scopes=["https://www.googleapis.com/auth/cloud-platform"])
```

However I didn't figure out a way to generate access token from "google.oauth2.service_account.Credentials"

**Comment by rmgogogo** [comment]
Here it is:

```
from google.oauth2 import service_account
import kfp

credentials = service_account.Credentials.from_service_account_file("/usr/local/google/home/renming/adc.json", scopes=["https://www.googleapis.com/auth/cloud-platform"])
token = credentials.token

client=kfp.Client(host='39e0f88a26ceb5d1-dot-us-central2.pipelines.googleusercontent.com', existing_token=token)
client.list_pipelines()
```

**Comment by rmgogogo** [comment]
I think we can close this ticket.

**Comment by amygdala** [comment]
great, I will give it a try!  Thank you!  With the original FR I was wondering whether we could push this handling into the SDK (it would probably require setting an env var or something).  what do you think? Could we leave this open if that seems feasible?  @rmgogogo

**Comment by amygdala** [comment]
> Here it is:
> 
> ```
> from google.oauth2 import service_account
> import kfp
> 
> credentials = service_account.Credentials.from_service_account_file("/usr/local/google/home/renming/adc.json", scopes=["https://www.googleapis.com/auth/cloud-platform"])
> token = credentials.token
> 
> client=kfp.Client(host='39e0f88a26ceb5d1-dot-us-central2.pipelines.googleusercontent.com', existing_token=token)
> client.list_pipelines()
> ```

Hmm, this isn't working for me.  I think it's important to test without gcloud being installed.  'token' is empty when I run the above code, and when I try to create the client, I get the 'gcloud not found' error.

/reopen

**Comment by aaaaahaaaaa** [comment]
@rmgogogo  Same for me, `credentials.token` is empty.

**Comment by rmgogogo** [comment]
Strange, how about these lines of codes in your side? It explicitly refresh the token first.

```
from google.oauth2 import service_account
from google.auth.transport import requests
import kfp

credentials = service_account.Credentials.from_service_account_file("/usr/local/google/home/renming/adc.json", scopes=["https://www.googleapis.com/auth/cloud-platform"])

credentials.refresh(requests.Request())
token = credentials.token

client=kfp.Client(host='39e0f88a26ceb5d1-dot-us-central2.pipelines.googleusercontent.com', existing_token=token)
client.list_pipelines()

```

**Comment by rmgogogo** [comment]
> Strange, how about these lines of codes in your side? It explicitly refresh the token first.
> 
> ```
> from google.oauth2 import service_account
> from google.auth.transport import requests
> import kfp
> 
> credentials = service_account.Credentials.from_service_account_file("/usr/local/google/home/renming/adc.json", scopes=["https://www.googleapis.com/auth/cloud-platform"])
> 
> credentials.refresh(requests.Request())
> token = credentials.token
> 
> client=kfp.Client(host='39e0f88a26ceb5d1-dot-us-central2.pipelines.googleusercontent.com', existing_token=token)
> client.list_pipelines()
> ```

I run these locally and it can work. Let me know if it still can't in your side.

**Comment by aaaaahaaaaa** [comment]
@rmgogogo Yes, that works! Thanks a lot, that solves it for me.

**Comment by amygdala** [comment]
yes!  that works for me too.  Thank you for figuring that out.

**Comment by amygdala** [comment]
(Let's keep this FR open _if_ you think it's a good idea to push some of this logic into the KF SDK so that there's less user-facing code needed. Perhaps this could be done via the user setting an env var).

**Comment by juliusvonkohout** [comment]
/close

Since there is no activity for 4 years
